### PR TITLE
Linq

### PR DESCRIPTION
**Pull request by vacho87**
To distinguish my comments (questions) in code comments // were made in russian language

### CODE REVIEW

File: LINQ.Exercises/Partition.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            // Здесь нагородил чего-то и оно как-то работает, НО
+            // 1. Приведение интерфейсного свойства PartitionNumbers к типу int [] это экспромт (типа методом тыка) и правильно ли это
+            // 2. Метод IndexOf() возвращает первое вхождение значения переменной number в массив а не индекс конкретного элемента,
+            // на который ссылается переменная number. То есть если бы источник данных выглядел бы, например, следующим образом:
+            // new int[] { 5, 4, 4, 4, 4, 4, 4, 1, 3, 9, 8, 6, 7, 2, 0 }; , то мой запрос бы дал неверный результат, так как каждый раз 
+            // условие отбора ложно удовлетворялось бы и в результирующую выборку попали бы все четверки, хотя для четверки с нидексом 5 
+            // фактически условие не выполняется.
+            // Можно пример правильного решения?
+            IEnumerable<int> result = TestData.PartitionNumbers.TakeWhile(number => number >= System.Array.IndexOf(TestData.PartitionNumbers as int [], number));
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Partition.cs:68]: `IEnumerable<int> result = TestData.PartitionNumbers.TakeWhile((number, index) => number >= index);`

File: LINQ.Exercises/Partition.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        // get the elements of the array
+        // starting from the first element less than its position.
+        [TestMethod]
+        public void GetElementsStartingFromFirstElementLessThanItsPosition_Return8ints()
+        {
+            // я все же нашел какой-то корявенький спопосб получить индекс переменной number внутри массива
+            int index = 0;
+            IEnumerable<int> result = TestData.PartitionNumbers.SkipWhile(number => number >= index++);
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Partition.cs:90]: `IEnumerable<int> result = TestData.PartitionNumbers.SkipWhile((number, index) => number >= index);`

File: LINQ.Exercises/Orderby.cs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            //
+            // 2. Так и не понял, что предпочтительней - наследование класса от Comparer, или реализация одного из интерфейсов
+            // (если релизация интерфейсов, то что предпочтительней Icomparer или Icomparable)? 
+            // Точнее мне не понятно почему предпочтительнее наследование нежели реализация. Все обсуждения этого вопроса на форумах, что я прочитал,
+            // цитируют все тот же MSDN, после которого я на эти форумы полез.
+            // Не мог бы ты мне пояснить почему именно наследование (в каких ситуациях это выгодно), и если наследование предпочтительнее, 
+            // то как быть в случаях, когда мне нужно наследовать свой класс от другого класса (естественно не Comparer),
+            // но в то же время необходимо реализовать функционал сравнения, ведь множественного наследования нет в C#.
+            // И если все же выбыирать реализацию интерфейса (опять же, когда это выгодно или необходимо?), то какой из них в каких случаях применять? 
+            var result = words.OrderBy(word => word);
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Orderby.cs:113]: please implement your own `Comparer<string>`

File: LINQ.Exercises/Orderby.cs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        {
+            #region hint
+
+            // you will have to create a class which implements the IComparer<string> interface
+
+            #endregion hint
+
+            string[] words = { "aPPLE", "AbAcUs", "bRaNcH", "BlUeBeRrY", "ClOvEr", "cHeRry" };
+
+            IEnumerable<string> result = words.OrderBy(word => word.Length).ThenBy(word => word);
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Orderby.cs:200]: please implement your own `Comparer<string>`

File: LINQ.Exercises/Orderby.cs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        }
+
+        // use Reverse to create a list of all digits in the array whose
+        // second letter is 'i' that is reversed from the order in the original array.
+        [TestMethod]
+        public void ComplexQuery_returnStringEnumeration()
+        {
+            string[] digits = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
+
+            IEnumerable<string> result = digits.Reverse().Where(word => word[1] == 'i');
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Orderby.cs:212]: always perform filtering before collection modification to modify smaller collection. in this case if you Reverse after Where Reverse will be faster as there are fewer elements in the sequence.

File: LINQ.Exercises/Element.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            int result = TestData.Numbers.Last(n => n % 2 == 0);
+
+            Assert.AreEqual(-4, result);
+        }
+
+        [TestMethod]
+        public void First_n_greater_than_10_if_not_found_return_0()
+        {
+            int result = TestData.Numbers.FirstOrDefault(n => n > 10);  // Заменил метод First на метод FirstOrDefault (или нужно было иначе?)
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Element.cs:58]: that's ok

File: LINQ.Exercises/CombinedOperations.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    /// These tests require the combination of various linq methods to arrive at a solution.
+    /// Hints as to what these methods are will not be given. The student is required to think 
+    /// critically above the problem in hand, and to utilise the full gamut of all the methods
+    /// available to him or her.
+    /// </summary>
+
+    // Почему-то методы этого класса не отображаются в ТестЭксплорере
+    // но в случае копирования в другой класс, например в SetOperations все исправно работает
+    [TestClass]
+    class CombinedOperations
```
Comment (alex-al-skiba) [inline LINQ.Exercises/CombinedOperations.cs]: it should be `public` to be visible in the explorer

File: LINQ.Exercises/Aggregate.cs [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        public void Find_latest_birthday()
+        {
+
+            // Первоначально строка ниже выглядела следующим образом: 
+            // "DateTime result = TestData.People.First().Born;"
+            // как и в некороых других упражнениях выше я ее и зменил принципиально и получил правильный результат
+            // но у меня периодически возникает вопрос:
+            // "а вдруг автор упражнения хотел чтобы упражнение было решено именно в рамках заданного им шаблона?"
+            // а я хоть и доился нужного результата, но добился его не тем путем
+            DateTime result = TestData.People.Max(person => person.Born);
```
Comment (alex-al-skiba) [inline LINQ.Exercises/Aggregate.cs:115]: looks correct

**Review by alex-al-skiba — CHANGES_REQUESTED** [review]

File: LINQ.Exercises/Orderby.cs (hunk 5 above)
Comment (vacho87) [reply LINQ.Exercises/Orderby.cs:212]: Roger

File: LINQ.Exercises/Orderby.cs (hunk 4 above)
Comment (vacho87) [reply LINQ.Exercises/Orderby.cs:200]: Will be done

File: LINQ.Exercises/Orderby.cs (hunk 3 above)
Comment (vacho87) [reply LINQ.Exercises/Orderby.cs:113]: Will be done

**Pull request merged**